File: python/venepagos/exceptions.py
"""Excepciones personalizadas del SDK de VenePagos."""

from typing import Any, Optional


class VenePagosError(Exception):
    """Error base del SDK de VenePagos."""

    def __init__(self, message: str) -> None:
        super().__init__(message)
        self.message = message


class AuthenticationError(VenePagosError):
    """Error de autenticacion: API key invalida o ausente."""

    def __init__(self, message: str = "API key invalida o no proporcionada.") -> None:
        super().__init__(message)


class APIError(VenePagosError):
    """Error devuelto por la API de VenePagos."""

    def __init__(
        self,
        message: str,
        status_code: int,
        body: Optional[Any] = None,
    ) -> None:
        super().__init__(message)
        self.status_code = status_code
        self.body = body

    def __str__(self) -> str:
        return f"APIError({self.status_code}): {self.message}"

File: dart/lib/src/exceptions.dart
/// Excepcion base para todos los errores del SDK de VenePagos.
class VenePagosException implements Exception {
  /// Mensaje descriptivo del error.
  final String message;

  /// Codigo de estado HTTP, si aplica.
  final int? statusCode;

  const VenePagosException(this.message, {this.statusCode});

  @override
  String toString() => 'VenePagosException: $message'
      '${statusCode != null ? ' (HTTP $statusCode)' : ''}';
}

/// Error de autenticacion (API key invalida o ausente).
class AuthenticationException extends VenePagosException {
  const AuthenticationException([String message = 'API key invalida o ausente'])
      : super(message, statusCode: 401);

  @override
  String toString() => 'AuthenticationException: $message';
}

/// Error generico de la API de VenePagos.
class ApiException extends VenePagosException {
  /// Cuerpo de la respuesta del servidor.
  final String? responseBody;

  const ApiException(
    super.message, {
    super.statusCode,
    this.responseBody,
  });

  @override
  String toString() => 'ApiException: $message'
      '${statusCode != null ? ' (HTTP $statusCode)' : ''}'
      '${responseBody != null ? '\nResponse: $responseBody' : ''}';
}
